File: CPP/ByTier/F/2162_F_Beautiful_Intervals.cpp
// Problem: CF 2162 F - Beautiful Intervals
// https://codeforces.com/contest/2162/problem/F

/*
Purpose: This code solves the problem of constructing a permutation of integers from 0 to n-1 such that the MEX (minimum excluded value) of all mex values computed over given intervals is minimized.

Algorithm:
- The solution attempts to construct a permutation that minimizes the MEX of the set of MEX values from all intervals.
- It uses a greedy approach to determine the best positions for elements 0, 1 to minimize the MEX.
- Special cases are handled for edge configurations like when an interval covers the full range [1,n].

Time Complexity: O(m log m + n^2) where m is the number of intervals and n is the size of permutation.
Space Complexity: O(n + m) for storing intervals and auxiliary arrays.

Techniques:
- Sorting and unique operations on intervals.
- Prefix sum technique to track interval coverage.
- Greedy placement of elements based on interval overlaps.
*/

#include <stdio.h>
#include <algorithm>
#include <utility>

const int maxn = 3e3 + 5;
using namespace std;
int T, n, m, cnt[maxn], ans[maxn], k;
pair<int, int> a[maxn];

// Function to build a permutation with specific elements at positions x and y
void build(int x, int y) {
  ans[x] = 0;
  ans[y] = 1;
  int tot = 2;
  for (int i = 1; i <= n; i++)
    if (i != x && i != y)
      ans[i] = tot++;
}

// Main solving logic
void solve() {
  // Initialize permutation as identity
  for (int i = 1; i <= n; i++)
    ans[i] = i - 1;
  
  // Try to find a position that is covered by all intervals
  for (int i = 1; i <= n; i++) {
    if (cnt[i] == k) {
      swap(ans[1], ans[i]);
      return;
    }
  }
  
  // Try to find a position not covered by any interval
  for (int i = 1; i <= n; i++) {
    if (!cnt[i]) {
      swap(ans[1], ans[i]);
      return;
    }
  }
  
  // Try to find two adjacent positions that are not covered by any interval
  for (int i = 1; i < n; i++) {
    int flag = 1;
    for (int j = 1; j <= k; j++) {
      if (i < a[j].first || a[j].second < i)
        continue;
      if (i + 1 < a[j].first || a[j].second < i + 1)
        flag = 0;
    }
    if (flag) {
      build(i, i + 1);
      return;
    }
  }
  
  // Try reverse adjacent check
  for (int i = 2; i <= n; i++) {
    int flag = 1;
    for (int j = 1; j <= k; j++) {
      if (i < a[j].first || a[j].second < i)
        continue;
      if (i - 1 < a[j].first || a[j].second < i - 1)
        flag = 0;
    }
    if (flag) {
      build(i, i - 1);
      return;
    }
  }
  
  // Fallback case where none of above conditions are met
  ans[1] = 0;
  ans[2] = 2;
  ans[3] = 1;
  for (int i = 4; i <= n; i++)
    ans[i] = i - 1;
}

int main() {
  scanf("%d", &T);
  for (; T; T--) {
    scanf("%d %d", &n, &m);
    k = 0;
    for (int i = 1; i <= n; i++)
      cnt[i] = 0;
    
    // Read intervals, skip full range [1,n] intervals
    for (int i = 1, l, r; i <= m; i++) {
      scanf("%d %d", &l, &r);
      if (l == 1 && r == n)
        continue;
      a[++k] = {l, r};
    }
    
    // Sort and remove duplicates
    sort(a + 1, a + k + 1);
    k = unique(a + 1, a + k + 1) - a - 1;
    
    // Compute prefix sum of interval coverage
    for (int i = 1; i <= k; i++)
      cnt[a[i].first]++, cnt[a[i].second + 1]--;
    for (int i = 1; i <= n; i++)
      cnt[i] += cnt[i - 1];
    
    solve();
    for (int i = 1; i <= n; i++)
      printf("%d ", ans[i]);
    puts("");
  }
  return 0;
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByRound/2162/2162_F_Beautiful_Intervals.py
# Problem: CF 2162 F - Beautiful Intervals
# https://codeforces.com/contest/2162/problem/F

"""
Algorithm: Beautiful Intervals
Purpose: To find a permutation of integers 0 to n-1 such that the mex of all mex values of intervals is minimized.
Techniques:
- Difference array technique for counting interval overlaps
- Greedy rearrangement of elements to minimize mex of result set M

Time Complexity: O(n * m) due to constructing difference arrays and iterating through intervals.
Space Complexity: O(n) for storing difference arrays and the answer permutation.

The approach uses prefix sum technique (difference array) to count how many intervals cover each position.
Then, we try different permutations based on overlap conditions to minimize mex(M).
"""

import sys

input = sys.stdin.readline


def readInt():
    return int(input())


def readArray(type=int, container=list):
    return container(map(type, input().split()))


def readBasic():
    return readInt(), readArray()


import itertools

for t in range(readInt()):
    n, m = readArray()
    # pref[i] stores how many intervals contain position i
    pref = [0] * (n + 1)
    # two[i] stores overlap for intervals of length > 1 at position i
    two = [0] * (n + 1)
    # other[i] stores overlap information for intervals starting from i
    other = [0] * (n + 1)
    
    # Process intervals using difference array
    for _ in range(m):
        l, r = readArray()
        pref[l - 1] += 1
        pref[r] -= 1
        if l != r:
            two[l - 1] += 1
            two[r - 1] -= 1
            other[l] += 1
            other[r] -= 1
    
    # Compute prefix sums for the difference arrays
    pref = list(itertools.accumulate(pref))
    two = list(itertools.accumulate(two))
    other = list(itertools.accumulate(other))
    
    # Initialize default answer as [1,2,3,...,n-1,0]
    ans = [(i + 1) % n for i in range(n)]
    
    # Try shifting array to reduce the mex based on overlap conditions
    for i in range(n - 1):
        # If number of intervals covering position i equals total intervals passing through it
        if two[i] == pref[i]:
            ans = [(j - i + n) % n for j in range(n)]
            break
    
    # Try further shifts
    for i in range(1, n):
        if other[i] == pref[i]:
            ans = [(i - j + n) % n for j in range(n)]
            break
    
    # If all positions are covered by same number of intervals
    for i in range(n):
        if pref[i] == m:
            ans = [(j - i + n) % n for j in range(n)]
            break
    
    print(" ".join(map(str, ans)))


# https://github.com/VaHiX/CodeForces/